test/utils/utils.go

```diff
@@ -0,0 +1,422 @@
+	}
+	if uid != 0 || gid != 0 {
+		nsEnterOpts := append([]string{"--userspec", fmt.Sprintf("%d:%d", uid, gid), "/", p.PodmanBinary}, podmanOptions...)
+		command = exec.Command("chroot", nsEnterOpts...)
+	} else {
+		command = exec.Command(p.PodmanBinary, podmanOptions...)
+	}
+	if env != nil {
+		command.Env = env
+	}
+
+	session, err := gexec.Start(command, GinkgoWriter, GinkgoWriter)
+	if err != nil {
+		Fail(fmt.Sprintf("unable to run podman command: %s\n%v", strings.Join(podmanOptions, " "), err))
+	}
+	return &PodmanSession{session}
+}
+
+// PodmanBase exec podman with default env.
+func (p *PodmanTest) PodmanBase(args []string) *PodmanSession {
+	return p.PodmanAsUser(args, 0, 0, nil)
+}
+
+// WaitForContainer waits on a started container
+func (p *PodmanTest) WaitForContainer() bool {
+	for i := 0; i < 10; i++ {
+		if p.NumberOfContainersRunning() > 0 {
+			return true
+		}
+		time.Sleep(1 * time.Second)
+	}
+	return false
+}
+
+// NumberOfContainersRunning returns an int of how many
+// containers are currently running.
+func (p *PodmanTest) NumberOfContainersRunning() int {
+	var containers []string
+	ps := p.PodmanBase([]string{"ps", "-q"})
+	ps.WaitWithDefaultTimeout()
+	Expect(ps.ExitCode()).To(Equal(0))
+	for _, i := range ps.OutputToStringArray() {
+		if i != "" {
+			containers = append(containers, i)
+		}
+	}
+	return len(containers)
+}
+
+// NumberOfContainers returns an int of how many
+// containers are currently defined.
+func (p *PodmanTest) NumberOfContainers() int {
+	var containers []string
+	ps := p.PodmanBase([]string{"ps", "-aq"})
+	ps.WaitWithDefaultTimeout()
+	Expect(ps.ExitCode()).To(Equal(0))
+	for _, i := range ps.OutputToStringArray() {
+		if i != "" {
+			containers = append(containers, i)
+		}
+	}
+	return len(containers)
+}
+
+// NumberOfPods returns an int of how many
+// pods are currently defined.
+func (p *PodmanTest) NumberOfPods() int {
+	var pods []string
+	ps := p.PodmanBase([]string{"pod", "ps", "-q"})
+	ps.WaitWithDefaultTimeout()
+	Expect(ps.ExitCode()).To(Equal(0))
+	for _, i := range ps.OutputToStringArray() {
+		if i != "" {
+			pods = append(pods, i)
+		}
+	}
+	return len(pods)
+}
+
+// GetContainerStatus returns the containers state.
+// This function assumes only one container is active.
+func (p *PodmanTest) GetContainerStatus() string {
+	var podmanArgs = []string{"ps"}
+	podmanArgs = append(podmanArgs, "--all", "--format={{.Status}}")
+	session := p.PodmanBase(podmanArgs)
+	session.WaitWithDefaultTimeout()
+	return session.OutputToString()
+}
+
+// WaitContainerReady waits process or service inside container start, and ready to be used.
+func (p *PodmanTest) WaitContainerReady(id string, expStr string, timeout int, step int) bool {
+	startTime := time.Now()
+	s := p.PodmanBase([]string{"logs", id})
+	s.WaitWithDefaultTimeout()
+
+	for {
+		if time.Since(startTime) >= time.Duration(timeout)*time.Second {
+			return false
```

Comment:
Is it possible within the test framework to display something here and other places where we've an unexpected error?  Just an error message like "WaitContainerReady timed out unexpectedly."  Ditto other places in this code.

Reply:
Hi @TomSweeneyRedHat Actually I prefer to just return true or false in the common functions. And left the test to decide if this is an expect results or not. Because in some tests, especially for some negative test, we expect the container will never in a ready status. The common functions will only provide method but not the judgement.

Reply:
@ypu that's fine.  If you're comfortable with that approach, that works for me.  I just thought some output there might help diagnose issues when  they arose..

Reply:
Another perspective related to what @TomSweeneyRedHat suggested:  Only returning True/False assumes that is the _only_ useful information.  The function cannot guarantee the state will not change between returning and gathering debugging details after-the-fact.  This can really make debugging maddening/crazy because you are hiding important information.  For example:

1. `WaitContainerReady()`  times out, returns false
1. container becomes ready
2. Test prints error message, including details about container.
3. details show: container is ready
4. engineer goes crazy

Since you can return multiple values in Go, why not use this and allow the caller to decide what is important.  For example, maybe return True/False, along with the actual/detected state of the container (command exit code, stdout/stderr, etc.)?  This way the behavior of the function will match the debugging details, even if something changes later.  For example:

1. `WaitContainerReady()` times out, returns false + container-details
1. container becomes ready
2. Test prints error message and matching container-details.
3. "ready" container can be ignored, test is already failed.
4. engineer can debug the real problem (timeout too short?)

Reply:
Hi @cevich 
Function Podman will print output to stdout directly. So we can get the details without return them as a part of the return value. And as we already can get the container status details, I prefer not return or print them again.

Reply:
> Function Podman will print output to stdout directly

Ohhh okay, yep that addresses my concern as well.  Thanks.